price-feeder/oracle/provider/mexc.go

```diff
@@ -103,35 +99,44 @@ func NewMexcProvider(
 		Path:   mexcWSPath,
 	}
 
-	wsConn, resp, err := websocket.DefaultDialer.Dial(wsURL.String(), nil)
-	defer resp.Body.Close()
-	if err != nil {
-		return nil, fmt.Errorf(
-			types.ErrWebsocketDial.Error(),
-			ProviderMexc,
-			err,
-		)
-	}
+	mexcLogger := logger.With().Str("provider", "mexc").Logger()
 
 	provider := &MexcProvider{
-		wsURL:           wsURL,
-		wsClient:        wsConn,
-		logger:          logger.With().Str("provider", "mexc").Logger(),
+		logger:          mexcLogger,
 		endpoints:       endpoints,
 		tickers:         map[string]types.TickerPrice{},
 		candles:         map[string][]types.CandlePrice{},
 		subscribedPairs: map[string]types.CurrencyPair{},
 	}
 
-	if err := provider.SubscribeCurrencyPairs(pairs...); err != nil {
-		return nil, err
-	}
+	provider.setSubscribedPairs(pairs...)
 
-	go provider.handleWebSocketMsgs(ctx)
+	controller := NewWebsocketController(
+		ctx,
+		ProviderMexc,
+		wsURL,
+		provider.getSubscriptionMsgs(),
+		provider.messageReceived,
+		mexcLogger,
+	)
+	go controller.Start()
```

Comment:
Why not use a pointer to the `provider` object instead?
```suggestion
	controller := NewWebsocketController(
		ctx,
		wsURL,
		ProviderMexc,
		*provider,
	)
	go controller.Start()
```

Reply:
maybe you could solve this with hooks? if it is not a common function for all providers

Reply:
`type MessageHandler func(int, []byte)` already acts like a hook. Do you have an example of what kind of hooks implementation you mean?

Reply:
yeap, like that